src/core/hooks/use-formts/use-formts.ts

```diff
@@ -46,47 +59,135 @@ export type FormtsOptions<Values extends object, Err> = {
 
 type FormtsReturn<Values extends object, Err> = [
   fields: FieldHandleSchema<Values, Err>,
-  form: Partial<FormHandle<Values, Err>> // TODO
+  form: FormHandle<Values, Err>
 ];
 
 export const useFormts = <Values extends object, Err>(
   options: FormtsOptions<Values, Err>
 ): FormtsReturn<Values, Err> => {
   /// INTERNAL STATE
   const [state, dispatch] = React.useReducer(
-    createReducer<Values>(),
+    createReducer<Values, Err>(),
     options,
     getInitialState
   );
 
   /// INTERNAL HANDLERS
-  const getField = <T, Err>(field: FieldDescriptor<T, Err>): T =>
+  const getField = <T>(field: FieldDescriptor<T, Err>): T =>
     get(state.values, impl(field).path) as any;
 
-  const isTouched = <T, Err>(field: FieldDescriptor<T, Err>) =>
+  const getFieldError = (field: FieldDescriptor<any, Err>): Err | null => {
+    const error = state.errors[impl(field).path];
+    return error == null ? null : error;
+  };
+
+  const isFieldTouched = <T>(field: FieldDescriptor<T, Err>) =>
     resolveTouched(get(state.touched as object, impl(field).path));
 
-  const touchField = <T, Err>(field: FieldDescriptor<T, Err>) =>
-    dispatch({ type: "touchValue", payload: { path: impl(field).path } });
+  const isFieldValid = <T>(field: FieldDescriptor<T, Err>) =>
+    resolveIsValid(state.errors, impl(field).path);
 
-  const setField = <T, Err>(field: FieldDescriptor<T, Err>, value: T): void => {
-    const decodeResult = impl(field).decode(value);
-    if (decodeResult.ok) {
-      dispatch({
-        type: "setValue",
-        payload: { path: impl(field).path, value: decodeResult.value },
+  const validateField = <T>(
+    field: FieldDescriptor<T, Err>,
+    trigger?: ValidationTrigger
+  ): Promise<void> => {
+    if (!options.validator) {
+      return Promise.resolve();
+    }
+
+    dispatch({ type: "setIsValidating", payload: { isValidating: true } });
+    return options.validator
+      .validate([field], getField, trigger)
+      .then(errors => {
+        setFieldErrors(...errors);
+        dispatch({ type: "setIsValidating", payload: { isValidating: false } });
       });
-    } else {
+  };
+
+  const validateForm = (): Promise<ValidationResult<Err>> => {
+    if (options.validator == null) {
+      return Promise.resolve([]);
+    }
+    const topLevelDescriptors = values(fields).map(it => it.descriptor);
+
+    dispatch({ type: "setIsValidating", payload: { isValidating: true } });
+    return options.validator
+      .validate(topLevelDescriptors, getField)
```

Comment:
this assumes validator runs validation recursively for nested fields when receiving root object or root array description. Looks like we don't have it implemented yet on the validator side - we can add this later after flattening the schema shape (should make it easier to implement)